File: sakinah-android/core/src/main/kotlin/org/emdatra/sakinah/core/ICS.kt
package org.emdatra.sakinah.core

import java.time.Instant
import java.time.ZoneOffset
import java.time.format.DateTimeFormatter

/** تصدير المواقيت إلى تقويم iCalendar (ICS) — مطابق لنسختي الويب وiOS */
object ICS {
  private val stamp: DateTimeFormatter = DateTimeFormatter.ofPattern("yyyyMMdd'T'HHmmss'Z'").withZone(ZoneOffset.UTC)
  fun date(i: Instant): String = stamp.format(i)
  fun escape(s: String): String = s.replace("\\", "\\\\").replace(";", "\\;").replace(",", "\\,").replace("\n", "\\n")

  /** طيّ السطور الأطول من 75 بايت (RFC 5545) دون قطع حرف متعدد البايتات */
  fun fold(line: String): String {
    val out = StringBuilder()
    var cur = StringBuilder()
    var bytes = 0
    for (ch in line) {
      val b = ch.toString().toByteArray(Charsets.UTF_8).size
      if (bytes + b > 75) { if (out.isNotEmpty()) out.append("\r\n"); out.append(cur); cur = StringBuilder(" "); bytes = 1 }
      cur.append(ch); bytes += b
    }
    if (out.isNotEmpty()) out.append("\r\n")
    out.append(cur)
    return out.toString()
  }

  data class Options(
    val locationName: String? = null,
    val prayers: Set<String> = Prayer.entries.map { it.id }.toSet(),
    val preMinutes: Int = 0,
    val includeSunrise: Boolean = false,
  )

  /** أيام من PrayerTimes.monthTable */
  fun build(days: List<PrayerTimesResult>, options: Options = Options(), now: Instant = Instant.now()): String {
    val lines = mutableListOf(
      "BEGIN:VCALENDAR", "VERSION:2.0", "PRODID:-//Sakinah//Prayer Times//AR",
      "CALSCALE:GREGORIAN", "METHOD:PUBLISH", "X-WR-CALNAME:مواقيت الصلاة — سكينة",
    )
    val ts = date(now)
    for (day in days) {
      for (key in Prayer.entries) {
        if (key.id !in options.prayers) continue
        if (key == Prayer.SUNRISE && !options.includeSunrise) continue
        val t = day[key] ?: continue
        val name = if (key == Prayer.SUNRISE) "الشروق" else "صلاة ${key.nameAr}"
        lines += listOf(
          "BEGIN:VEVENT", "UID:${date(t)}-${key.id}@sakinah", "DTSTAMP:$ts",
          "DTSTART:${date(t)}", "DTEND:${date(t.plusSeconds(600))}",
          "SUMMARY:${escape(name)}",
          "DESCRIPTION:${escape(name + (options.locationName?.let { " — $it" } ?: "") + "\nمن تطبيق سكينة")}",
          "TRANSP:TRANSPARENT",
        )
        if (key != Prayer.SUNRISE) {
          lines += listOf("BEGIN:VALARM", "ACTION:DISPLAY", "DESCRIPTION:${escape(name)}", "TRIGGER:PT0M", "END:VALARM")
          if (options.preMinutes > 0) {
            lines += listOf("BEGIN:VALARM", "ACTION:DISPLAY", "DESCRIPTION:${escape("اقترب موعد $name")}", "TRIGGER:-PT${options.preMinutes}M", "END:VALARM")
          }
        }
        lines += "END:VEVENT"
      }
    }
    lines += "END:VCALENDAR"
    return lines.joinToString("\r\n") { fold(it) } + "\r\n"
  }
}

File: sakinah-native/Sources/SakinahCore/ICS.swift
import Foundation

/// تصدير المواقيت إلى تقويم iCalendar (ICS) — مطابق لـ buildICS في notifications.js
public enum ICS {
  static let stampFormatter: DateFormatter = { let f = DateFormatter(); f.locale = Locale(identifier: "en_US_POSIX"); f.timeZone = TimeZone(identifier: "UTC"); f.dateFormat = "yyyyMMdd'T'HHmmss'Z'"; return f }()
  public static func date(_ d: Date) -> String { stampFormatter.string(from: d) }
  public static func escape(_ s: String) -> String {
    s.replacingOccurrences(of: "\\", with: "\\\\").replacingOccurrences(of: ";", with: "\\;").replacingOccurrences(of: ",", with: "\\,").replacingOccurrences(of: "\n", with: "\\n")
  }
  /// طيّ السطور الأطول من 75 بايت (RFC 5545) دون قطع حرف متعدد البايتات
  public static func fold(_ line: String) -> String {
    var out: [String] = []; var cur = ""; var bytes = 0
    for ch in line {
      let b = ch.utf8.count
      if bytes + b > 75 { out.append(cur); cur = " "; bytes = 1 }
      cur.append(ch); bytes += b
    }
    out.append(cur); return out.joined(separator: "\r\n")
  }
  public struct Options: Sendable { public var locationName: String? = nil; public var prayers: Set<Prayer> = Set(Prayer.allCases); public var preMinutes: Int = 0; public var includeSunrise = false; public init() {} }

  /// أيام من PrayerTimes.compute
  public static func build(days: [PrayerTimesResult], options: Options = Options(), now: Date = Date()) -> String {
    var lines = ["BEGIN:VCALENDAR", "VERSION:2.0", "PRODID:-//Sakinah//Prayer Times//AR", "CALSCALE:GREGORIAN", "METHOD:PUBLISH", "X-WR-CALNAME:مواقيت الصلاة — سكينة"]
    let stamp = date(now)
    for day in days {
      for key in Prayer.allCases {
        guard options.prayers.contains(key) else { continue }
        if key == .sunrise && !options.includeSunrise { continue }
        guard let t = day[key] else { continue }
        let name = key == .sunrise ? "الشروق" : "صلاة \(key.nameAr)"
        let uid = "\(date(t))-\(key.rawValue)@sakinah"
        lines += ["BEGIN:VEVENT", "UID:\(uid)", "DTSTAMP:\(stamp)", "DTSTART:\(date(t))", "DTEND:\(date(t.addingTimeInterval(600)))",
                  "SUMMARY:\(escape(name))", "DESCRIPTION:\(escape("\(name)\(options.locationName.map { " — \($0)" } ?? "")\nمن تطبيق سكينة"))", "TRANSP:TRANSPARENT"]
        if key != .sunrise {
          lines += ["BEGIN:VALARM", "ACTION:DISPLAY", "DESCRIPTION:\(escape(name))", "TRIGGER:PT0M", "END:VALARM"]
          if options.preMinutes > 0 { lines += ["BEGIN:VALARM", "ACTION:DISPLAY", "DESCRIPTION:\(escape("اقترب موعد \(name)"))", "TRIGGER:-PT\(options.preMinutes)M", "END:VALARM"] }
        }
        lines.append("END:VEVENT")
      }
    }
    lines.append("END:VCALENDAR")
    return lines.map(fold).joined(separator: "\r\n") + "\r\n"
  }
}
